design/specs/clipboard-interactions.md

```diff
@@ -0,0 +1,315 @@
+
+- When the paste data has an empty string for a cell value, `null` is used if possible, with an empty string as a fallback for non-nullable text columns.
+
+- When multiple cells are selected during paste:
+
+    - The paste data is repeated as needed to completely fill the selection. (Same as **Calc**.)
+    - If the paste data extends beyond the selection, then the selection is increased to fit the paste data. (Same as **Sheets**.)
+
+#### Pasting cells into columns of different types
+
+1. Column metadata from the source data is matched to columns in the target table. For each pairing, the following logic takes place:
+    - Some pairings of column types cannot be pasted. For example an `email` column cannot be pasted into a `number` column. These pairings result in an error.
+    - If the target column is a text column and the source column has formatted values, then Mathesar asks the user what to do. Examples of this include copying FK data into a text column, or copying dates into a text column.
+
+1. If any pairing results in an error, Mathesar aborts the paste and displays a toast error to the user such as:
+
+    > **Unable to Paste**
+    >
+    > Email data cannot be pasted into a Number column.
+
+1. If any pairing results in a "ask" outcome, then Mathesar prompts the user via a dialog like the following:
+
+    > **Paste**
+    >
+    > The data you are pasting can be represented in multiple ways. Paste the raw values to use the text
+    >
+    > `Paste Raw Values` &nbsp; &nbsp; &nbsp; `Paste Formatted Values`
+
+    "Paste Formatted Values" is selected by default.
+
+    When "Paste Formatted Values" is chosen, formatted values will be used for the paste when possible. Otherwise, raw values will be used.
+
+
+## Design imperfections and future considerations
+
+- In my [initial "copy" implementation](https://github.com/centerofci/mathesar/pull/2773), I notice that copying a lot of cells takes some time and **locks the main thread.** This is not great, but it's because we're using the sync API. It's hard to get around this though since the async API isn't fully implemented in Firefox yet. We could potentially use the async API if it's available, while falling back to the sync API for Firefox. If we want to do that, I'd rather do it later when/if users ask for improved performance here.
+
+- It would be nice to add **context menu entries for copy/paste**, but there are some challenges here.
+
+    - When copying: If we use the synchronous clipboard API it seems[^1] that we can't trigger copy events programmatically. Somehow Google Sheets is doing it, but I haven't dug deeper to figure out how. With the async clipboard API Firefox lacks support for MIME types other than `text/plain`.
+
+    - When pasting: The synchronous clipboard API doesn't allow programmatic pasting at all. The async clipboard API doesn't work allow pasting in Firefox and requires a special permission in other browsers.
```

Comment:
Doesn't the clipboard API support this?

https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/writeText

We could use `navigator.clipboard.writeText()` method to write to the clipboard, the only caveat being Firefox only supports this event from a user action, which will be the case for context menus.

For paste, we'd be reading from the clipboard using the `navigator.clipboard.readText()` method, and performing "paste" manually i.e. updating the cells ourselves.

I remember doing this a long time ago and there were some challenges like the elements need to be in focus. 

I'm adding this comment here, perhaps we could update the document with this as well, so that when we want to implement this, this comment can be a starting point.

Reply:
It does. Perhaps for Firefox we can limit support to `text/plain` and try working with that, and/or we could write items to localstorage as a way to improve copy/paste support within Mathesar.

Either way, I agree this isn't a high priority at the moment, and we can consider picking it up later.